File: frontend/src/types/contracts.ts
export type RunMode = 'demo' | 'live';
export type ProductMode = 'search' | 'committee';
export type RunStage =
  | 'idle'
  | 'analyzing'
  | 'ready'
  | 'starting'
  | 'running'
  | 'stopping'
  | 'completed'
  | 'error';

export type PersonaState =
  | 'idle'
  | 'searching'
  | 'success'
  | 'partial'
  | 'failure'
  | 'reacting';

export type ExperimentDecision = 'kept' | 'reverted';

export interface LlmConfig {
  provider: 'openai_compatible' | 'openai' | 'anthropic' | 'disabled';
  baseUrl?: string;
  model?: string;
  apiKey?: string;
}

export interface LexicalFieldEntry {
  field: string;
  boost: number;
}

export interface ConnectRequest {
  mode: RunMode;
  esUrl?: string;
  apiKey?: string;
  indexName?: string;
  llm?: LlmConfig;
  uploadedEvalSet?: EvalCase[];
  autoGenerateEval?: boolean;
  vectorFieldOverride?: string | null;
  maxSampleDocs?: number;
}

export interface ConnectResponse {
  connectionId: string;
  productMode: ProductMode;
  mode: RunMode;
  stage: RunStage;
  summary: ConnectionSummary;
  warnings: string[];
}

export interface ConnectionSummary {
  clusterName: string;
  clusterVersion?: string;
  indexName: string;
  docCount: number;
  detectedDomain: 'security' | 'developer_docs' | 'compliance' | 'general';
  primaryTextFields: string[];
  vectorField?: string | null;
  vectorDims?: number | null;
  sampleDocs: SampleDoc[];
  baselineEvalCount: number;
  baselineReady: boolean;
}

export interface SampleDoc {
  id: string;
  title: string;
  excerpt: string;
  fieldPreview: Record<string, string>;
}

export interface EvalCase {
  id: string;
  query: string;
  relevantDocIds: string[];
  personaHint?: string;
  difficulty?: 'easy' | 'medium' | 'hard';
}

export interface SearchProfile {
  lexicalFields: LexicalFieldEntry[];
  multiMatchType: 'best_fields' | 'most_fields' | 'cross_fields' | 'phrase';
  minimumShouldMatch: string;
  tieBreaker: number;
  phraseBoost: number;
  fuzziness: 'AUTO' | '0';
  useVector: boolean;
  vectorField?: string | null;
  vectorWeight: number;
  lexicalWeight: number;
  fusionMethod: 'weighted_sum' | 'rrf';
  rrfRankConstant: number;
  knnK: number;
  numCandidates: number;
}

export interface SearchProfileChange {
  path: string;
  before: string | number | boolean | null;
  after: string | number | boolean | null;
  label: string;
}

export interface ExperimentRecord {
  experimentId: number;
  timestamp: string;
  hypothesis: string;
  change: SearchProfileChange;
  beforeScore: number;
  baselineScore?: number;
  candidateScore: number;
  deltaAbsolute: number;
  deltaPercent: number;
  decision: ExperimentDecision;
  durationMs: number;
  queryFailuresBefore: string[];
  queryFailuresAfter: string[];
}

export interface PersonaDefinition {
  id: string;
  name: string;
  role: string;
  department: string;
  archetype: string;
  goal: string;
  orbit: number;
  colorSeed: number;
  queries: string[];
}

export interface PersonaRuntime {
  id: string;
  state: PersonaState;
  lastQuery?: string;
  lastResultRank?: number | null;
  successRate: number;
  totalSearches: number;
  successes: number;
  partials: number;
  failures: number;
  angle: number;
  speed: number;
  radius: number;
  pulseUntil?: number | null;
  reactUntil?: number | null;
}

export interface PersonaViewModel extends PersonaDefinition, PersonaRuntime {}

export interface PersonaActivityEntry {
  id: string;
  kind: 'query' | 'success' | 'partial' | 'failure' | 'reacting';
  title: string;
  detail: string;
  timestamp: string;
}

export interface RunLaunchConfig {
  durationMinutes: number;
  maxExperiments: number;
  personaCount: number;
  autoStopOnPlateau: boolean;
}

export interface CompressionMethodResult {
  method: 'float32' | 'int8' | 'int4' | 'rotated_int4';
  sizeBytes: number;
  recallAt10: number;
  estimatedMonthlyCostUsd: number;
  sizeReductionPct: number;
  status: 'pending' | 'running' | 'done' | 'skipped' | 'error';
  note?: string;
}

export interface CompressionSummary {
  available: boolean;
  vectorField?: string | null;
  vectorDims?: number | null;
  methods: CompressionMethodResult[];
  bestRecommendation?: string;
  projectedMonthlySavingsUsd?: number;
  status: 'idle' | 'running' | 'done' | 'skipped' | 'error';
}

export interface HeroMetrics {
  currentScore: number;
  baselineScore: number;
  bestScore: number;
  improvementPct: number;
  experimentsRun: number;
  improvementsKept: number;
  personaSuccessRate: number;
  elapsedSeconds: number;
  projectedMonthlySavingsUsd?: number | null;
  scoreTimeline: Array<{ t: number; score: number }>;
  // Continuation tracking — cumulative progress across run chain
  originalBaselineScore?: number | null;
  priorExperimentsRun?: number;
  priorImprovementsKept?: number;
}

export interface RunConfig {
  durationMinutes: number;
  maxExperiments: number;
  personaCount: number;
  autoStopOnPlateau: boolean;
}

export interface RunSnapshot {
  runId: string;
  productMode: ProductMode;
  mode: RunMode;
  stage: RunStage;
  summary: ConnectionSummary;
  searchProfile: SearchProfile;
  recommendedProfile: SearchProfile;
  metrics: HeroMetrics;
  personas: PersonaViewModel[];
  experiments: ExperimentRecord[];
  compression: CompressionSummary;
  warnings: string[];
  runConfig?: RunConfig;
  startedAt?: string;
  completedAt?: string | null;
}

export interface QueryBreakdownRow {
  baselineTopResults?: Array<{
    docId: string;
    title: string;
    excerpt: string;
    score: number;
  }>;
  bestTopResults?: Array<{
    docId: string;
    title: string;
    excerpt: string;
    score: number;
  }>;
  queryId: string;
  query: string;
  difficulty: string;
  baselineScore: number;
  bestScore: number;
  deltaPct: number;
  failureReason?: string | null;
  topRelevantDocIds: string[];
}

export interface PersonaImpactRow {
  personaId: string;
  name: string;
  role: string;
  beforeSuccessRate: number;
  afterSuccessRate: number;
  deltaPct: number;
}

export interface ReportPayload {
  runId: string;
  generatedAt: string;
  mode: RunMode;
  summary: {
    headline: string;
    overview: string;
    nextSteps: string[];
    baselineScore: number;
    bestScore: number;
    improvementPct: number;
    experimentsRun: number;
    improvementsKept: number;
    durationSeconds: number;
    projectedMonthlySavingsUsd?: number | null;
    // Continuation tracking
    isContinuation?: boolean;
    originalBaselineScore?: number | null;
    totalExperimentsRun?: number | null;
    totalImprovementsKept?: number | null;
  };
  connection: ConnectionSummary;
  connectionConfig?: {
    mode: RunMode;
    esUrl?: string | null;
    apiKey?: string | null;
    hasApiKey?: boolean;
    indexName?: string | null;
    evalSet: EvalCase[];
    llm?: LlmConfig | null;
  } | null;
  searchProfileBefore: SearchProfile;
  searchProfileAfter: SearchProfile;
  diff: SearchProfileChange[];
  queryBreakdown: QueryBreakdownRow[];
  personaImpact: PersonaImpactRow[];
  experiments: ExperimentRecord[];
  compression: CompressionSummary;
  warnings: string[];
  previousRunId?: string | null;
}

export interface SearchRunListItem {
  run_id: string;
  mode: RunMode;
  stage: RunStage;
  index_name?: string | null;
  cluster_name?: string | null;
  baseline_score: number;
  best_score: number;
  improvement_pct: number;
  experiments_run: number;
  started_at?: string | null;
  completed_at?: string | null;
  updated_at?: string | null;
}

export interface BenchmarkHealthPreset {
  id: string;
  label: string;
  indexName: string;
  expectedDocCount: number;
  docCount: number;
  ready: boolean;
  setupCommand: string;
  reachable: boolean;
}

export interface QueryPreviewPayload {
  queryId: string;
  query: string;
  baselineResults: Array<{
    docId: string;
    title: string;
    excerpt: string;
    score: number;
  }>;
  optimizedResults: Array<{
    docId: string;
    title: string;
    excerpt: string;
    score: number;
  }>;
  baselineQueryDsl?: Record<string, unknown> | null;
  optimizedQueryDsl?: Record<string, unknown> | null;
}

export type RunSocketEvent =
  | { type: 'snapshot'; payload: RunSnapshot }
  | { type: 'run.stage'; payload: { runId: string; stage: RunStage; message?: string } }
  | { type: 'run.complete'; payload: { runId: string; stage: RunStage } }
  | { type: 'experiment.completed'; payload: ExperimentRecord }
  | { type: 'persona.batch'; payload: { runId: string; personas: PersonaRuntime[] } }
  | { type: 'metrics.tick'; payload: HeroMetrics }
  | { type: 'compression.updated'; payload: CompressionSummary }
  | { type: 'report.ready'; payload: ReportPayload }
  | { type: 'error'; payload: { code: string; message: string } };

// ── Codex branch additions ─────────────────────────────────────────────────
export interface ModelComparisonEntry {
  modelId: string;
  baselineScore: number;
  bestScore: number;
  improvementPct: number;
  experimentsRun: number;
  improvementsKept: number;
  bestProfile: SearchProfile;
  topChanges: string[];
}

export interface ModelComparisonResult {
  entries: ModelComparisonEntry[];
  recommendedModel: string | null;
  comparisonNote: string;
}

export interface ReportNarrativeSection {
  key: string;
  title: string;
  body: string;
  audience?: 'executive' | 'operator' | 'technical';
  source?: 'deterministic' | 'llm';
  confidence?: number | null;
}

export interface ReportPersonaSummary {
  personaCount: number;
  archetypeCounts: Record<string, number>;
  topRoles: string[];
  explanation: string;
}

export interface ReportValidationNote {
  title: string;
  body: string;
  severity: 'success' | 'info' | 'warning';
  confidence?: number | null;
}

export interface ReportSnippetLine {
  lineNumber: number;
  content: string;
  changed?: boolean;
  explanation?: string | null;
}

export interface ReportCodeSnippet {
  title: string;
  target: string;
  format: string;
  summary: string;
  beforeLines: ReportSnippetLine[];
  afterLines: ReportSnippetLine[];
}

export interface ReportChangeNarrative {
  path: string;
  title: string;
  plainEnglish: string;
  before: string;
  after: string;
  expectedEffect: string;
  whyItHelped: string;
  confidence?: number | null;
  evidence: string[];
}

export interface ReportImplementationGuide {
  summary: string;
  applyInstructions: string[];
  representativeQuery?: string | null;
  note?: string | null;
  snippets: ReportCodeSnippet[];
}

File: backend/models/contracts.py
from __future__ import annotations
from pydantic import BaseModel, Field
from typing import Optional, List, Literal, Any, Dict
from enum import Enum

RunMode = Literal['demo', 'live']
ProductMode = Literal['search', 'committee']
RunStage = Literal['idle', 'analyzing', 'ready', 'starting', 'running', 'stopping', 'completed', 'error']
PersonaState = Literal['idle', 'searching', 'success', 'partial', 'failure', 'reacting']
ExperimentDecision = Literal['kept', 'reverted']


class LlmConfig(BaseModel):
    provider: Literal['openai_compatible', 'openai', 'anthropic', 'disabled'] = 'disabled'
    baseUrl: Optional[str] = None
    model: Optional[str] = None
    apiKey: Optional[str] = None


class EvalCase(BaseModel):
    id: str
    query: str
    relevantDocIds: List[str]
    personaHint: Optional[str] = None
    difficulty: Optional[Literal['easy', 'medium', 'hard']] = None


class ConnectRequest(BaseModel):
    mode: RunMode
    esUrl: Optional[str] = None
    apiKey: Optional[str] = None
    indexName: Optional[str] = None
    llm: Optional[LlmConfig] = None
    uploadedEvalSet: Optional[List[EvalCase]] = None
    autoGenerateEval: bool = True
    vectorFieldOverride: Optional[str] = None
    maxSampleDocs: int = 120


class SampleDoc(BaseModel):
    id: str
    title: str
    excerpt: str
    fieldPreview: Dict[str, str]


class ConnectionSummary(BaseModel):
    clusterName: str
    clusterVersion: Optional[str] = None
    indexName: str
    docCount: int
    detectedDomain: Literal['security', 'developer_docs', 'compliance', 'general']
    primaryTextFields: List[str]
    vectorField: Optional[str] = None
    vectorDims: Optional[int] = None
    sampleDocs: List[SampleDoc] = []
    baselineEvalCount: int = 0
    baselineReady: bool = False


class ConnectResponse(BaseModel):
    connectionId: str
    productMode: ProductMode = 'search'
    mode: RunMode
    stage: RunStage
    summary: ConnectionSummary
    warnings: List[str] = []


class SearchProfile(BaseModel):
    lexicalFields: List[Dict[str, Any]] = []
    multiMatchType: Literal['best_fields', 'most_fields', 'cross_fields', 'phrase'] = 'best_fields'
    minimumShouldMatch: str = '75%'
    tieBreaker: float = 0.0
    phraseBoost: float = 0.0
    fuzziness: Literal['AUTO', '0'] = '0'
    useVector: bool = False
    vectorField: Optional[str] = None
    vectorWeight: float = 0.35
    lexicalWeight: float = 0.65
    fusionMethod: Literal['weighted_sum', 'rrf'] = 'weighted_sum'
    rrfRankConstant: int = 60
    knnK: int = 20
    numCandidates: int = 100


class SearchProfileChange(BaseModel):
    path: str
    before: Any
    after: Any
    label: str


class ExperimentRecord(BaseModel):
    experimentId: int
    timestamp: str
    hypothesis: str
    change: SearchProfileChange
    baselineScore: float
    candidateScore: float
    deltaAbsolute: float
    deltaPercent: float
    decision: ExperimentDecision
    durationMs: int
    queryFailuresBefore: List[str] = []
    queryFailuresAfter: List[str] = []


class PersonaDefinition(BaseModel):
    id: str
    name: str
    role: str
    department: str
    archetype: str
    goal: str
    orbit: int
    colorSeed: int
    queries: List[str]


class PersonaRuntime(BaseModel):
    id: str
    state: PersonaState = 'idle'
    lastQuery: Optional[str] = None
    lastResultRank: Optional[int] = None
    successRate: float = 0.0
    totalSearches: int = 0
    successes: int = 0
    partials: int = 0
    failures: int = 0
    angle: float = 0.0
    speed: float = 0.08
    radius: float = 120.0
    pulseUntil: Optional[float] = None
    reactUntil: Optional[float] = None


class PersonaViewModel(PersonaDefinition, PersonaRuntime):
    pass


class CompressionMethodResult(BaseModel):
    method: Literal['float32', 'int8', 'int4', 'rotated_int4']
    sizeBytes: int
    recallAt10: float
    estimatedMonthlyCostUsd: float
    sizeReductionPct: float
    status: Literal['pending', 'running', 'done', 'skipped', 'error'] = 'pending'
    note: Optional[str] = None


class CompressionSummary(BaseModel):
    available: bool = False
    vectorField: Optional[str] = None
    vectorDims: Optional[int] = None
    methods: List[CompressionMethodResult] = []
    bestRecommendation: Optional[str] = None
    projectedMonthlySavingsUsd: Optional[float] = None
    status: Literal['idle', 'running', 'done', 'skipped', 'error'] = 'idle'


class HeroMetrics(BaseModel):
    currentScore: float = 0.0
    baselineScore: float = 0.0
    bestScore: float = 0.0
    improvementPct: float = 0.0
    experimentsRun: int = 0
    improvementsKept: int = 0
    personaSuccessRate: float = 0.0
    elapsedSeconds: float = 0.0
    projectedMonthlySavingsUsd: Optional[float] = None
    scoreTimeline: List[Dict[str, float]] = []


class RunConfig(BaseModel):
    durationMinutes: int = 30
    maxExperiments: int = 200
    personaCount: int = 36
    autoStopOnPlateau: bool = True


class RunSnapshot(BaseModel):
    runId: str
    productMode: ProductMode = 'search'
    mode: RunMode
    stage: RunStage
    summary: ConnectionSummary
    searchProfile: SearchProfile
    recommendedProfile: SearchProfile
    metrics: HeroMetrics
    personas: List[PersonaViewModel] = []
    experiments: List[ExperimentRecord] = []
    compression: CompressionSummary
    warnings: List[str] = []
    runConfig: RunConfig = Field(default_factory=RunConfig)
    startedAt: Optional[str] = None
    completedAt: Optional[str] = None


class StartRunRequest(BaseModel):
    connectionId: str
    durationMinutes: int = 30
    maxExperiments: int = 200
    personaCount: int = 36
    autoStopOnPlateau: bool = True
    previousRunId: Optional[str] = None  # Resume from a previous run's best profile


class StartRunResponse(BaseModel):
    runId: str
    productMode: ProductMode = 'search'
    stage: RunStage


class StopRunResponse(BaseModel):
    runId: str
    productMode: ProductMode = 'search'
    stage: RunStage


# WebSocket event union type aliases (used as dicts in practice)
